File: client/src/types/Activity.ts
export type Activity = {
  id: number
  title: string
  totalAmount: number
  date: string
}

export type ActivityCreated = {
  title: string
  date: string
}

export type Transaction = {
  id: number
  category: {
    id: number
    name: string
  }
  payer: {
    id: number
    name: string
  }
  date: string
  amount: number
  paid: boolean
  description: string
}

export type TransactionItemResult = {
  id: number
  category: {
    id: string
    name: string
  }
  date: string
  amount: number
  paid: boolean
  description: string
  activity: {
    id: number
    title: string
  }
}

export type TransactionCreated = {
  categoryId: number
  date: string
  amount: number
  paid: boolean
  description: string
  payerId: number
}

export type TransactionEdited = TransactionCreated & {
  id: number
}

File: server/MoneyFlies.Api/Entities/Activity.cs
using System.Collections.ObjectModel;

namespace MoneyFlies.Api.Entities;

public class Activity
{
    private Activity()
    {

    }

    public Activity(string title, DateOnly date)
    {
        Title = title;
        Date = date;
        TotalAmount = 0;
        Transactions = [];
    }

    public int Id { get; private set; }
    public string Title { get; private set; }
    public DateOnly Date { get; private set; }
    public decimal TotalAmount { get; private set; }
    public Collection<Transaction> Transactions { get; private set; } = [];

    public Transaction AddTransaction(
        Category category, string description, decimal amount, bool paid, DateOnly date, Payer payer)
    {
        Transaction transaction = new(this, category, description, amount, paid, date, payer);
        Transactions.Add(transaction);
        TotalAmount += amount;
        return transaction;
    }

    public Transaction UpdateTransaction(
        int transactionId, Category category, string description, decimal amount, bool paid, DateOnly date, Payer payer)
    {
        Transaction? transaction = Transactions.SingleOrDefault(t => t.Id == transactionId) ?? throw new ArgumentException("Transaction not found");
        TotalAmount -= transaction.Amount;
        TotalAmount += amount;

        transaction.Update(category, description, amount, paid, date, payer);

        return transaction;
    }

    public void RemoveTransaction(Transaction transaction)
    {
        Transactions.Remove(transaction);
        TotalAmount -= transaction.Amount;
    }
}
